Draft: Blockwise scaling linear quantization recipe

kwyss-nvidia opened the pull request:
# Description

Add a recipe configuration to quantizers and GEMM for float8 blockwise scaling.

Quantizer MR: https://github.com/NVIDIA/TransformerEngine/pull/1513
GEMM MR: https://github.com/NVIDIA/TransformerEngine/pull/1545

Pending before removing draft status:

Golden tensor dumps for exact linear/layer norm testing.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Add recipe and recipe state classes to create quantizers.
- Modify linear, layer norm linear, and layer_norm_mlp to use the quantized data paths.
- Test for linear and layer_norm.
- Revert recent change that breaks test_numerics for all recipes.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

kwyss-nvidia commented:
@ptrendx here is a parallel review from my fork with only the recipe changes.
https://github.com/kwyss-nvidia/TransformerEngine/pull/2

kwyss-nvidia commented:
@zhongbozhu did you check in tensor dumps anywhere? I ran against goldens, but don't see a tensor_dumps folder in TransformerEngine. Did we decide against checking them in?

yaox12 commented on transformer_engine/pytorch/fp8.py:
```diff
@@ -900,3 +930,94 @@ def make_quantizers(self) -> list:
+                    force_pow_2_scales=self.recipe.fp8_quant_fwd_inp.power_2_scale,
+                    block_scaling_dim=self.recipe.x_block_scaling_dim,
+                ),
+            ] * (self.num_quantizers // 3)
```
We can't use `[] * ` here. In this way, the same **object** will be repeated.
Properly should be 
```python
            return list(itertools.chain.from_iterable([[
                Float8BlockQuantizer(
                    fp8_dtype=self.qx_dtype,
                    rowwise=True,
                    columnwise=True,
                    amax_epsilon=self.recipe.fp8_quant_fwd_inp.amax_epsilon,
                    force_pow_2_scales=self.recipe.fp8_quant_fwd_inp.power_2_scale,
                    block_scaling_dim=self.recipe.x_block_scaling_dim,
                ),
                Float8BlockQuantizer(
                    fp8_dtype=self.qx_dtype,
                    rowwise=True,
                    columnwise=True,
                    amax_epsilon=self.recipe.fp8_quant_fwd_weight.amax_epsilon,
                    force_pow_2_scales=self.recipe.fp8_quant_fwd_weight.power_2_scale,
                    block_scaling_dim=self.recipe.w_block_scaling_dim,
                ),
                Float8BlockQuantizer(
                    fp8_dtype=self.qx_dtype,
                    rowwise=True,
                    columnwise=True,
                    amax_epsilon=self.recipe.fp8_quant_fwd_inp.amax_epsilon,
                    force_pow_2_scales=self.recipe.fp8_quant_fwd_inp.power_2_scale,
                    block_scaling_dim=self.recipe.x_block_scaling_dim,
                ),
            ] for _ in range(self.num_quantizers // 3)]))
```

kwyss-nvidia replied:
Thanks for taking a look. The aliasing was obscuring a configuration bug. I've applied your feedback and fixed the bug.

yaox12 commented on transformer_engine/pytorch/fp8.py:
```diff
@@ -900,3 +930,94 @@ def make_quantizers(self) -> list:
+                force_pow_2_scales=self.recipe.fp8_quant_bwd_grad.power_2_scale,
+                block_scaling_dim=self.recipe.grad_block_scaling_dim,
+            ),
+        ] * (self.num_quantizers // 2)
```
Ditto.

kwyss-nvidia replied:
Done.

BestJuly commented on transformer_engine/pytorch/tensor/_internal/float8_blockwise_tensor_base.py:
```diff
@@ -0,0 +1,240 @@
+    ) -> Tuple[list[Optional[torch.Tensor]], Float8BlockwiseQTensorBase]:
+        """Prepare the tensor base for saving for backward"""
+        tensors = [self._rowwise_data, self._columnwise_data]
+        self._rowwise_data = None
```
This will raise a bug when PP is enabled. Files related to other recipes have been fixed in [PR1548](https://github.com/NVIDIA/TransformerEngine/pull/1548).

kwyss-nvidia replied:
Done. Thanks for finding this issue. I would like to investigate whether the rowwise data is actually cleared when the forward pass is complete based on a call to update the usage. It would be nice to be aggressive about memory optimization.

kwyss-nvidia commented on tests/pytorch/distributed/test_numerics.py:
```diff
@@ -48,12 +48,14 @@ def _run_test(quantization):
     if quantization == "mxfp8" and not mxfp8_available:
         pytest.skip(reason_for_no_mxfp8)
+    if quantization == "fp8_block_scaling" and not fp8_block_scaling_available:
+        pytest.skip(reason_for_no_fp8_block_scaling)
```
Good catch. Now in this file as well as the other `test_numerics.py`.

zhongbozhu commented on transformer_engine/pytorch/module/linear.py:
```diff
@@ -144,7 +146,7 @@ def forward(
                 if not isinstance(inputmat, QuantizedTensor):
                     columnwise_usage = backward_needs_input and isinstance(
-                        input_quantizer, MXFP8Quantizer
+                        input_quantizer, (MXFP8Quantizer, Float8BlockQuantizer)
```
Let me dump our offline discussion here. As we discussed offline, the logic here is quite tricky when dealing with subchannel blockwise recipe. 

Back in FP8, we only have one scaling factors per-tensor, so it's okay for us to set columnwise usage to False and rowwise usage to be True. But now since the scaling block for X is 1x128, and in backward we need 128x1. We actually need to run a quantize & transpose kernel. However, saving transposed output for backward will make it challenging to gather using Dim 0. 

At this point, I think it's okay for us to just save bf16 X for backward at the price of increasing GPU memory usage and this shall serve as a baseline implementation. In longer term, we need to implement a bidirectional quantization kernel which calculates 1x128 and 128x1 quantization result but don't do any transpose. So then we can save the untransposed qx for backward, and we can also gather it again in backward with Dim 0.

kwyss-nvidia replied:
As discussed, this is targeting BF16 gather.

kunlunl commented on transformer_engine/common/recipe/recipe_common.cuh:
```diff
@@ -32,18 +32,13 @@ __device__ __forceinline__ float compute_scale_from_amax(float amax, float max_f
-
-  if (isnan(scale)) {
-    scale = 1.f;
+    scale = inf_value;
```
Also, I think you need to add `std::numeric_limits<float>::max()` to https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/csrc/extensions/multi_tensor/multi_tensor_compute_scale.cu#L50-L51 since you added a `value_for_inf` argument.

kwyss-nvidia replied:
Done. Made a mistake locally compiling from the wrong branch and missed this error.

zhongbozhu commented on tests/cpp/test_common.cu:
```diff
@@ -259,27 +305,33 @@ Tensor::Tensor(const std::string& name,
       }
     }
+    if (q_opts != nullptr) {
+      tensor_.set_qopt_force_pow_2_scales(q_opts->force_pow_2_scales);
```
this API is now gone right? we might need to fix cpp tests

kwyss-nvidia replied:
These are fixed.

zhongbozhu commented on tests/pytorch/test_float8_blockwise_gemm_exact.py:
Root cause found: this test case assumes using pow2scale False => causing assertion failures in C++, so it will be fixed after changing the quantizer config. 

When we have more bandwidth, let's remove the assert and make pow2scales fully configurable => though this will change the nvte_quantize API.

kwyss-nvidia replied:
Updated the qa_pytorch_unittest script.

kwyss-nvidia replied:
GEMM test also passes after I updated power of 2 scales.

zhongbozhu commented on transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:
```diff
@@ -0,0 +1,597 @@
+        dst._fp8_dtype = self.dtype
+        return dst
+
+    def get_scale_shape(self, shape: Iterable[int], columnwise: bool) -> Tuple[int, int]:
```
As discussed offline, having scaling factors swizzled and permuted will make it harder to do FP8 communication, which a  pre-condition for optimizing this recipe with TPGEMM overlap. 

We can add one more option to this API with a default to False option called compact. 

Or we can actually make this compact scale property native to the quantizer, since every layer has its own quantizers, and each quantizer can be individually configured based on the communication pattern. Example here: https://github.com/NVIDIA/TransformerEngine/blob/160be219575d5fe3aa4913cb3e3c0cc7bee245f5/transformer_engine/pytorch/module/linear.py#L1214

So for row_parallel + sequence parallel, its fwd inp doesn't need allgather, so let's make quantization kernel fuse the swizzling. For column_parallel + sequence parallel, its fwd inp needs all gather, then we should make a compact scale tensor, and swizzle it before GEMM. 

That being said, let's do BF16 gather first so this is not a problem for now. But let's revisit it after we get the correct baseline. @timmoon10 @BestJuly .

kwyss-nvidia replied:
Resolving. Will revisit for FP8 gather, but not on this MR.

zhongbozhu commented on transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:
```diff
@@ -0,0 +1,597 @@
+        # pylint: disable=missing-function-docstring
+        return Float8BlockwiseQTensor.make_like(self)
+
+    def update_usage(
```
This function is likely not correct. It seems that the code of this function is similar with MXFP8 quantizer. However, the difference here is that MXFP8 assumes Blackwell tensor core's transpose capability. While this block scaling recipe is meant to be used for Hopper & Blackwell. 

If you look at the update_usage function of Float8Tensor, it creates transpose data if it's missing. And our linear layers rely on this feature to prepare data for GEMMs.

kwyss-nvidia commented on transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:
```diff
@@ -0,0 +1,597 @@
+            dst._fp8_dtype = src._fp8_dtype
+            dst._rowwise_scale_inv = src._rowwise_scale_inv
+            dst._columnwise_scale_inv = src._columnwise_scale_inv
+            if dst.requires_grad != src.requires_grad:
```
Removing this would be consistent with pytorch Tensor's data accessor, which I believe should be the goal. I will adopt your change and remove these lines, and I'm hopeful nothing will regress.

kwyss-nvidia replied:
Done.

zhongbozhu commented on transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:
```diff
@@ -0,0 +1,613 @@
+    # cleanly by stacking data without aliasing tiles and whether
+    # the scales also stack on the proper dimensions.
+
+    def make_empty(
```
Since it's possible that we only have columnwise usage = True with rowwise usage = False if we choose to do BF16 gather and save the ungathered bf16 tensor for backward. This make_empty function should also take this into account to skip the malloc of rowwise data.

zhongbozhu commented:
I added the support for doing columnwise only mode to the 1D block quantization kernel here: https://github.com/kwyss-nvidia/TransformerEngine/commit/31941880d5d0d2e19f6960d6748218f6030a632e

Since we are doing bf16 gather now, it's important that in bwd, we only quantize to the columnwise usage tensor. 

Please run the two tests here:
```
pytest tests/pytorch/test_float8blockwisetensor.py::TestFloat8BlockwiseTensor::test_quantize_dequantize_columnwise_only -s -v

pytest tests/pytorch/distributed/test_numerics.py -s -v
```

API is changed from `bool return_transpose` to the two usage options below, @timmoon10 @kwyss-nvidia  please share your thoughts about the API design here. Later when we move on to fp8 gather with swizzling turn on/off, we can build on top of this.
```
// enum class for rowwise usage: None, ROWWISE
// there is no need to transpose the rowwise usage tensor, so no ROWWISE_TRANSPOSE
enum class RowwiseUsageOption {
  NONE, // No rowwise data
  ROWWISE // Rowwise data
};

// enum class for columnwise usage: None, COLUMNWISE, COLUMNWISE_TRANSPOSE
// For Hopper sm90 with only TN fp8 gemm, there is need to do columnwise transpose when doing 1D block scaling
enum class ColumnwiseUsageOption {
  NONE, // No columnwise data
  COLUMNWISE, // TODO(zhongbo): Columnwise data (current not supported getting columnwise data without transposing)
  COLUMNWISE_TRANSPOSE // Columnwise data with transpose
};
```

kwyss-nvidia commented:
> I added the support for doing columnwise only mode to the 1D block quantization kernel here: [kwyss-nvidia@bf958b1](https://github.com/kwyss-nvidia/TransformerEngine/commit/bf958b16b45c63907074655c821f8d70800d4b73)
> 
> Since we are doing bf16 gather now, it's important that in bwd, we only quantize to the columnwise usage tensor.
> 
> Please run the two tests here:
> 
> ```
> pytest tests/pytorch/test_float8blockwisetensor.py::TestFloat8BlockwiseTensor::test_quantize_dequantize_columnwise_only -s -v
> 
> pytest tests/pytorch/distributed/test_numerics.py -s -v
> ```
> 
> API is changed from `bool return_transpose` to the two usage options below, @timmoon10 @kwyss-nvidia please share your thoughts about the API design here. Later when we move on to fp8 gather with swizzling turn on/off, we can build on top of this.
> 
> ```
> // enum class for rowwise usage: None, ROWWISE
> // there is no need to transpose the rowwise usage tensor, so no ROWWISE_TRANSPOSE
> enum class RowwiseUsageOption {
>   NONE, // No rowwise data
>   ROWWISE // Rowwise data
> };
> 
> // enum class for columnwise usage: None, COLUMNWISE, COLUMNWISE_TRANSPOSE
> // For Hopper sm90 with only TN fp8 gemm, there is need to do columnwise transpose when doing 1D block scaling
> enum class ColumnwiseUsageOption {
>   NONE, // No columnwise data
>   COLUMNWISE, // TODO(zhongbo): Columnwise data (current not supported getting columnwise data without transposing)
>   COLUMNWISE_TRANSPOSE // Columnwise data with transpose
> };
> ```

This is cherry picked. I moved the enum out of common into the header with the kernel launcher and removed the currently unsupported `COLUMNWISE` enumeration for now since it's unsupported.

The tests you mentioned pass.

zhongbozhu commented:
Please cherry-pick this commit here: https://github.com/kwyss-nvidia/TransformerEngine/commit/dc86fd55cb5373316a38bbee714ad8af11dcfcb8

This commit adds the new `nvte_quantize_v2` api with the extra slot of putting in advanced numerical configs like power_of_2_scales.

zhongbozhu commented on transformer_engine/pytorch/module/layernorm_linear.py:
```diff
@@ -170,10 +172,49 @@ def forward(
+        # so we need to set with_quantized_norm to False
+        if isinstance(input_quantizer, Float8CurrentScalingQuantizer):
+            with_quantized_norm = False
+        if isinstance(input_quantizer, Float8BlockQuantizer):
```
with the new `nvte_quantize_v2`, this check is not needed anymore and the quantizer kernel launches will be done in C++ (same thing for current scaling)

zhongbozhu commented on transformer_engine/pytorch/module/layernorm_mlp.py:
```diff
@@ -214,6 +216,9 @@ def forward(
         with_quantized_norm = (
             fp8 and not return_layernorm_output and not return_layernorm_output_gathered
         )
+        if isinstance(fc1_input_quantizer, Float8BlockQuantizer):
```
like above, with `nvte_quantize_v2`, this check is not needed

zhongbozhu commented:
After adding `recipe.Float8BlockScaling()` to the recipe list of `test_cuda_graph.py`, the cuda graph test is going to fail with the `linear_op` case.

```
PYTORCH_JIT=0 NVTE_TORCH_COMPILE=0 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0 python3 -m pytest tests/pytorch/test_cuda_graphs.py::test_make_graphed_callables -s -v
```

If you add `recipe.Float8BlockScaling()` to the recipe list of `test_sanity.py`. Besides adding a pytest skip in the group linear test case `test_sanity_grouped_linear`, There is another error shown below.  

```
pytest tests/pytorch/test_sanity.py::test_sanity_linear_with_zero_tokens -s -v
```
This is due to the `_set_from_tensor` method of `Float8BlockwiseQTensor`. 
```
    def _set_from_tensor(dst: Float8BlockwiseQTensor, src: Float8BlockwiseQTensor):
        dst._rowwise_data = src._rowwise_data
        dst._columnwise_data = src._columnwise_data
        dst._quantizer = src._quantizer
        dst._fp8_dtype = src._fp8_dtype
        dst._rowwise_scale_inv = src._rowwise_scale_inv
        dst._columnwise_scale_inv = src._columnwise_scale_inv
>       dst.dtype = src.dtype
E       AttributeError: attribute 'dtype' of 'torch._C.TensorBase' objects is not writable

transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:485: AttributeError
```

kwyss-nvidia commented:
> After adding `recipe.Float8BlockScaling()` to the recipe list of `test_cuda_graph.py`, the cuda graph test is going to fail with the `linear_op` case.
> 
> ```
> PYTORCH_JIT=0 NVTE_TORCH_COMPILE=0 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0 python3 -m pytest tests/pytorch/test_cuda_graphs.py::test_make_graphed_callables -s -v
> ```
> 
> If you add `recipe.Float8BlockScaling()` to the recipe list of `test_sanity.py`. Besides adding a pytest skip in the group linear test case `test_sanity_grouped_linear`, There is another error shown below.
> 
> ```
> pytest tests/pytorch/test_sanity.py::test_sanity_linear_with_zero_tokens -s -v
> ```
> 
> This is due to the `_set_from_tensor` method of `Float8BlockwiseQTensor`.
> 
> ```
>     def _set_from_tensor(dst: Float8BlockwiseQTensor, src: Float8BlockwiseQTensor):
>         dst._rowwise_data = src._rowwise_data
>         dst._columnwise_data = src._columnwise_data
>         dst._quantizer = src._quantizer
>         dst._fp8_dtype = src._fp8_dtype
>         dst._rowwise_scale_inv = src._rowwise_scale_inv
>         dst._columnwise_scale_inv = src._columnwise_scale_inv
> >       dst.dtype = src.dtype
> E       AttributeError: attribute 'dtype' of 'torch._C.TensorBase' objects is not writable
> 
> transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:485: AttributeError
> ```

Removed setting the dtype in [this commit](https://github.com/NVIDIA/TransformerEngine/pull/1559/commits/80057a67a7418e203dec2bbb16d11a1aa4bc93da).

BestJuly commented on transformer_engine/pytorch/tensor/float8_blockwise_tensor.py:
I think it should be better to add `_data` in this file, as in TE2.x code there are many places using `tensor._data` to obtain the data.
```
    @property
    def _data(self) -> torch.Tensor:
        return self._rowwise_data if self._rowwise_data else self._columnwise_data
```

kwyss-nvidia replied:
In what scenarios is it helpful to access or manipulate the underlying data without the corresponding dequantization scales?

I can think of one scenario where the data tensors are saved and restored between fwd and backward and memory is reclaimed in the meantime.

The `_data` method, while inherited from torch, only imperfectly fits a class like this or `MXFP8Tensor` where there are multiple, numerically distinct data representations. Perhaps preferring `_rowwise_data` is a sensible convention to force a fit. It seems like callers will likely have to fiddle usage to have `_data` achieve a desired outcome.

BestJuly replied:
OK. I am suggesting this because there are some codes like `if ln_out._data is None` [code](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/module/layernorm_linear.py#L681). Then maybe we can add some `if` there to avoid `._data` method for blockwise recipe.

kwyss-nvidia replied:
That code should be updated when relevant user buffer support is added. Calling `_data` to check None can be made more clear/compatible.

kwyss-nvidia commented on transformer_engine/pytorch/fp8.py:
```diff
@@ -923,8 +953,121 @@ def __init__(
+                )
+                quantizers.append(
+                    Float8BlockQuantizer(
+                        fp8_dtype=self.qx_dtype,
```
Seems like this should be `self.qw_dtype`. Will fix.

kwyss-nvidia commented on transformer_engine/pytorch/fp8.py:
```diff
@@ -994,73 +993,81 @@ def __init__(
+            # Handle ops that do not support FP8 gemm output.
+            quantizers_per_gemm = 3 if fp8_output else 2
+            assert self.num_quantizers % quantizers_per_gemm == 0
+            num_gemms = self.num_quantizers // quantizers_per_gemm
```
Is the expectation of MOE that there are independent quantizers for each expert or do they reuse the same x and w quantizers?

zhongbozhu replied:
Maybe we should get rid of `quantizers_per_gemm` and just use:
```
num_gemms = self.num_quantizers // 3
```

kwyss-nvidia replied:
That's related to the CUDA_GRAPH commit.

zhongbozhu replied:
For MoE, each gemm (ie. each expert) has its own quantizers, this number of quantizers should be either 2 or 3 depending on fp8 output. @yaox12 can confirm. 

`m_splits` is the same with `num_gemms`, and `num_gemms` determines the number of quantizers. 

https://github.com/NVIDIA/TransformerEngine/blob/b362a6e071a8867d5d376351a43beb76d547e4e1/transformer_engine/pytorch/module/grouped_linear.py#L595

https://github.com/NVIDIA/TransformerEngine/blob/b362a6e071a8867d5d376351a43beb76d547e4e1/transformer_engine/pytorch/module/grouped_linear.py#L617

kwyss-nvidia replied:
The latest code should be compatible with that.

zhongbozhu commented on transformer_engine/pytorch/fp8.py:
```diff
@@ -994,73 +993,81 @@ def __init__(
+        assert self.mode == "backward", f"Unexpected mode {self.mode}"
+        quantizers_per_gemm = 2 if fp8_output else 1
+        assert self.num_quantizers % quantizers_per_gemm == 0
+        num_gemms = self.num_quantizers // quantizers_per_gemm
```
`num_gemms = self.num_quantizers // 2`

kwyss-nvidia commented on transformer_engine/pytorch/ops/op.py:
```diff
@@ -234,7 +234,7 @@ def _reset_quantization_recipe_state(
 
             # Construct builder class for quantized tensors
-            self._quantizers[mode] = recipe_state.make_quantizers()
+            self._quantizers[mode] = recipe_state.make_quantizers(fp8_output=False)
```
How do we more loudly fail if this codepath is applicable, but the caller has passed the `fp8_output` flag?

kwyss-nvidia replied:
@ksivaman

timmoon10 commented on transformer_engine/pytorch/module/linear.py:
```diff
@@ -609,6 +618,12 @@ def backward(ctx, grad_output: torch.Tensor) -> Tuple[Union[torch.Tensor, None],
+                    ):
+                        # Async gather in BF16 does not asynchronously
+                        # call quantizer after gather.
+                        inputmat_total = ctx.input_quantizer(inputmat_total)
```
Should we set the quantizer usage here?
```suggestion
                        ctx.input_quantizer.set_usages(rowwise=False, columnwise=True)
                        inputmat_total = ctx.input_quantizer(inputmat_total)
```

zhongbozhu replied:
There is no need to set it, the actual set_usage happens here in the `else` branch shared with mxfp8:
https://github.com/NVIDIA/TransformerEngine/blob/db2aaa9e7765a3bcd1e652e3539d57e85cdbd5bc/transformer_engine/pytorch/module/linear.py#L518

kwyss-nvidia replied:
It's more clear if the usage is set. Added those calls.

timmoon10 commented on transformer_engine/pytorch/module/layernorm_linear.py:
```diff
@@ -170,10 +172,49 @@ def forward(
+                if isinstance(input_quantizer, MXFP8Quantizer):
+                    with_quantized_norm = False
+
+        # Reduce duplicated transpose in `_fix_gathered_fp8_transpose`
+        if (
+            fp8
+            and FP8GlobalStateManager.get_fp8_recipe().float8_per_tensor_scaling()
+            and ub_bulk_dgrad
+        ):
+            input_quantizer.set_usage(rowwise=True, columnwise=False)
```
Why are we thinking about the dgrad GEMM in the forward pass? Should this be `ub_overlap_ag_fprop`?

zhongbozhu replied:
That was the issue we discussed with @BestJuly when working on tpgemm overlap. By default, the usage is both True, meaning that we will trigger quantize transpose kernel if we run `out = quantizer(inp)`. However, since we are doing tpgemm overlap, only one usage will be copied into ub, leading to a wasted transpose op and a wasted columnwise usage malloc. 

The `_fix_gathered_fp8_transpose` happens right after dgrad finishes because dgrad ub gemm will gather inputmat, and then we can transpose it. This means that if we don't have this branch, we will be doing two transpose ops.

zhongbozhu replied:
Actually, you might also be right. Maybe we should also add a or condition here for `ub_overlap_ag_fprop`? @BestJuly What do you think?

BestJuly replied:
I think we do not need these code in the fwd pass, same as @Tim's suggestion. This is because in current logic of main branch, transpose has been split from quantization in such case:
```
           if (
                columnwise_usage
                and with_input_all_gather
                and not isinstance(input_quantizer, MXFP8Quantizer)
            ):
                columnwise_usage = False
            input_quantizer.set_usage(rowwise=True, columnwise=columnwise_usage)
```
The transpose will not be included in the fwd pass and in the bwd, `_fix_gathered_fp8_transpose` will be called if `ub_bulk_dgrad=True`, and if not, `ln_out_total.update_usage(columnwise_usage=True)` will do the transpose.

In my mind the only issue would be 1. in previous code (prior to @Tim's PR), transposed will be conducted in the quantize kernel; 2. in current logic, we need to call two kernels, one in fwd pass which does not transpose data, the other one in backward pass which transpose data. Not sure whether this will introduce exposed CPU overhead to affect e2e perf.

kwyss-nvidia replied:
This forward pass is now simplified and distributed tests are still passing.

kwyss-nvidia commented on transformer_engine/common/recipe/__init__.py:
```diff
@@ -81,6 +81,10 @@ def float8_per_tensor_scaling(self):
         """Whether the given recipe is per-tensor scaling."""
         return isinstance(self, (DelayedScaling, Float8CurrentScaling))
 
+    def fp8blockwise(self):
```
Done.

kwyss-nvidia commented on transformer_engine/pytorch/fp8.py:
```diff
@@ -796,7 +826,7 @@ def create(
 
     @abc.abstractmethod
-    def make_quantizers(self) -> list:
+    def make_quantizers(self, fp8_output: bool = True) -> list:
```
After looking at the commit adding this and enabling CUDA_GRAPH testing, perhaps a simpler solution is to update num_quantizers in `basic_linear.py` to adhere to the convention of `x, w, output`.

The API change is removed and seeing passing results from:
```
PYTORCH_JIT=0 NVTE_TORCH_COMPILE=0 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0 pytest tests/pytorch/test_cuda_graphs.py
```

timmoon10 replied:
I think we should skip the CUDA graph test case with the linear op and sub-channel scaling. We would need to make more changes to support sub-channel scaling in `te.Sequential`.

kwyss-nvidia replied:
Done.

timmoon10 commented on transformer_engine/pytorch/fp8.py:
```diff
@@ -928,3 +959,108 @@ def make_quantizers(self) -> list:
+        from .tensor.float8_blockwise_tensor import Float8BlockQuantizer
+
+        if self.mode == "forward":
+            # The index convention (coming from base.py set_meta_tensor)
+            # is somewhat awkward, and doesn't play nicely with QuantizeOp,
+            # which is not associated with a GEMM.
+            assert self.num_quantizers % 3 == 0  # x, w, output per gemm
```
Hacky, but I think this is a reasonable approach for now.

The root cause is that `RecipeState` assumes you can just specify the number of quantizers. However, FP8 block scaling requires different kinds of quantizers for activations and weights. The modules can keep track of this by explicitly creating 3 quantizers per GEMM (input, weight, output), but this is not general and it results in unnecessary quantizers (for `LayerNormMLP` you don't need a quantizer for FC1 output). Perhaps we can modify `RecipeState` so you specify what kind of quantizers you need instead of just providing a number, but I worry about how to make this general and not over-tuned toward linear layers.

timmoon10 commented on transformer_engine/pytorch/ops/basic/basic_linear.py:
```diff
@@ -254,10 +255,16 @@ def _canonicalize_tensor_parallelism(
         )
 
     def num_quantizers(self, mode: str) -> int:
+        # Adhere to consistent conventions of the non-fused
+        # module code, where fwd quantized tensors are (x, w, output)
+        # and bwd quantized tensors are (ygrad, dgrad)
+        #
+        # Fused code need not use the output or dgrad quantizers.
+        # Since fp8_output flag is not available.
```
I don't think this is a good convention. It's specific to linear layers (it doesn't even really fit `LayerNormMLP`) and it doesn't match how the linear op handles FP8 output (it uses the next op's quantizer, which allows for more intelligent logic for quantizer usages).

For now, it would be better if we just didn't support sub-channel scaling with the linear op. We should skip the CUDA graph test case that started us down this path.

kwyss-nvidia replied:
OK. I've disabled cuda graph tests and fail out in `op.py` `_reset_quantization_recipe_state` for this recipe until the full CUDA graph design is addressed.

zhongbozhu replied:
will CUDA graph be enabled in actual E2E training? will this impact training latency?

kwyss-nvidia replied:
This recipe is specific to linear layers. I'm surprised there is utility to having cuda graphs and fusible ops if linear is not in scope.

If we update the num_quantizers, the cuda graph tests pass. Maybe we need a more principled solution than that to enable, but you mentioned you foresee additional changes for sequential. How do we scope this correctly?

timmoon10 replied:
The linear op is a separate code path from `te.Linear`/`LayerNormLinear`/`LayerNormMLP`. It's part of an experimental API for fusible operations, which is much more flexible and self-contained than our existing modules (think about the work required if we wanted a SwiGLULinear layer...). It's fine if that API doesn't support sub-channel scaling yet.

kwyss-nvidia replied:
Added the other cases back to the tests, but skip linear_op.

timmoon10 commented on transformer_engine/pytorch/distributed.py:
```diff
@@ -1075,7 +1153,13 @@ def gather_along_first_dim(
+    """
+    All-gather tensors and concatenate along first dimension.
+
+    NOTE: Caller should be aware that there are asynchronous cases
+    where quantizer is not None, but the output will not be quantized.
+    This affects Float8BlockQuantizer.
```
This makes it too hard to trust that `gather_along_first_dim` is doing what you expect. We should enforce a contract that `gather_along_first_dim` respects `quantizer`. I think this functional already does the right thing, so we just need to change the linear modules. For the forward pass, we just need to make sure that the input is a high-precision tensor and this should already do the right thing. For the backward pass, we should pass in `quantizer=None` so that we can have async communication.

kwyss-nvidia replied:
If the caller passes async_op and a quantizer in a scenario where there is not an implementation to honor both, should we fall back to synchronous or raise an error?

kwyss-nvidia replied:
Done. Callers are updated to pass in None for a quantizer in the relevant cases where async is desired, but async + quantize is unavailable.

timmoon10 commented on transformer_engine/pytorch/csrc/common.h:
```diff
@@ -187,6 +188,9 @@ class Float8BlockQuantizer : public Quantizer {
   // Gets rowwise and columnwise_data from tensor and sets them on wrapper
   void set_quantization_params(TensorWrapper* tensor) const override;
 
+  // Set options for quantization on QuantizationConfigWrapper.
+  void set_quantization_config(QuantizationConfigWrapper* quant_config) const;
```
This name makes me think we are setting the quantizer, which is opposite the actual behavior.
```suggestion
```
I don't think this function provides much value over setting the config directly.

kwyss-nvidia replied:
Removed and set directly.

timmoon10 commented on transformer_engine/pytorch/csrc/extensions/cast.cpp:
```diff
@@ -61,15 +64,20 @@ py::object quantize(const at::Tensor& tensor, py::handle quantizer, const py::ob
     // set amax ptr to null in te_output TensorWrapper to avoid atomic amax updates in kernel
     te_output.set_amax(nullptr, DType::kFloat32, te_output.defaultShape);
+  } else if (detail::IsFloat8BlockwiseQuantizers(quantizer.ptr())) {
+    auto my_quantizer_bw = static_cast<Float8BlockQuantizer*>(my_quantizer.get());
+    my_quantizer_bw->set_quantization_config(&quant_config);
```
I don't think `set_quantization_config` is providing much value over just setting the config object directly.
```suggestion
    auto my_quantizer_bw = static_cast<Float8BlockQuantizer*>(my_quantizer.get());
    quant_config.set_force_pow_2_scales(my_quantizer_bw->force_pow_2_scales);
    quant_config.set_amax_epsilon(my_quantizer_bw->amax_epsilon);
```
It's not ideal that we have to duplicate this in the norm functions, but the right approach is to refactor this code (in a future PR). Instead of calling `nvte_quantize_v2` directly here, each quantizer class should have a separate `quantize` function.

kwyss-nvidia replied:
Removed the method and use the structs.

zhongbozhu commented on transformer_engine/pytorch/module/layernorm_linear.py:
```diff
@@ -171,10 +168,23 @@ def forward(
                 columnwise_usage
                 and with_input_all_gather
                 and not isinstance(input_quantizer, MXFP8Quantizer)
+                and not isinstance(input_quantizer, Float8BlockQuantizer)
```
Why should we check for block quantizer here? 
MXFP8 checks for it so that when doing mxfp8, columnwise usage will be True => so that mxfp8 cast will be doing 1x32 and 32x1 without any transpose => then they do fp8 gather in rowwise usage => then they save columnwise usage for backward to gather again.

In our case, we are already targeting BF16 gather, meaning that should not do quantize transpose in fwd because we need to save bf16 unquantized tensor for backward.

kwyss-nvidia replied:
If we're not doing gather at all, we want columnwise_usage=True and we want to save the columnwise data to ctx. But you're right this line is not necessary. Deleted.

timmoon10 commented on transformer_engine/pytorch/module/linear.py:
```diff
@@ -130,6 +131,10 @@ def forward(
         )
         own_quantized_input = False
+        # TODO(kwyss): Support FP8 allgather for FP8 block quantization.
+        force_hp_input_gather = (
+            fp8 and with_input_all_gather_nccl and isinstance(input_quantizer, Float8BlockQuantizer)
+        )
```
Nit: "HP" is not very clear.
```suggestion
        force_hp_input_gather = (
            fp8 and with_input_all_gather_nccl and isinstance(input_quantizer, Float8BlockQuantizer)
        )  # Perform TP communication in high precision
```

kwyss-nvidia replied:
Done.

ptrendx commented on transformer_engine/common/include/transformer_engine/transformer_engine.h:
```diff
@@ -286,6 +286,12 @@ enum NVTEQuantizationConfigAttribute {
   kNVTEQuantizationConfigAmaxEpsilon = 1,
+  /*! Noop tensor (containing a scalar).
+   If the scalar element value = 1, quantization kernel will early exit.
+   This is a tensor in order that the flag can be on GPU and conditional
+   early exit is compatible with a static CUDA graph.
+  */
```
```suggestion
   This is a tensor because the flag must be on the GPU in order to enable
   conditional early exit even when captured in static CUDA graph.
  */
```

kwyss-nvidia replied:
Done.

ptrendx commented on transformer_engine/common/recipe/__init__.py:
```diff
@@ -287,3 +292,112 @@ def __post_init__(self) -> None:
+                    use for calculating dgrad in backward pass
+    fp8_gemm_wgrad: MMParams, default MMParams.use_split_accumulator=True
+                    use for calculating dgrad in backward pass
+    fp8_dpa: bool, default = `False`
```
Having this documented will be confusing to people since FP8 DPA is supported only for the delayed scaling recipe. Same with fp8_mha.

kwyss-nvidia replied:
Removed from doc.

timmoon10 commented on tests/pytorch/test_float8_current_scaling_exact.py:
```diff
@@ -82,7 +82,8 @@ def _get_sum_abs_error(a, b):
     @staticmethod
     def _get_mean_abs_relative_error(a, b):
-        return torch.mean(torch.abs((a - b) / b))
+        error = torch.where(b == 0, 0.0, torch.abs((a - b) / b))
+        return torch.mean(error)
```
How do we handle the case where `a != 0` and `b == 0`? Perhaps:
```suggestion
        error = torch.where(b == 0, torch.ne(a, b), torch.abs((a - b) / b))
        return torch.mean(error)
```

kwyss-nvidia replied:
The relative error is undefined for b==0. Arguably, a principled thing to do would be to mask those pixels out from the mean so the average is calculated over fewer elements.

Perhaps we can use mean from https://pytorch.org/docs/stable/masked.html

I have not used the masked APIs yet.

kwyss-nvidia replied:
Maybe a simple solution is to choose the non-zero entry from either a or b as the denominator when one is zero, and an arbitrary non-zero denominator if if they are both equal to zero:

```
denom = torch.where(b == 0, torch.where(a == 0, 1.0, a))
error = torch.abs((a - b) / denom)
```

timmoon10 replied:
Our implementations work out to the same thing, and whatever feels more intuitive is fine. We are basically assuming that the two values are on the same order of magnitude.

timmoon10 commented on tests/pytorch/test_float8_blockwise_scaling_exact.py:
```diff
@@ -199,12 +193,90 @@ def test_quantization_block_tiling_versus_reference(
     [
         # full tile cases
         (128, 128),
+        (256, 256),
+        (256, 1024),
+        (1024, 256),
+        # Padding required cases
+        (256, 272),
+        (303, 300),
+        (305, 256),
+        # Some larger tiles.
+        (2000, 2000),
+        (2048, 2000),
+        (2000, 1024),
+        (2048, 1024),
```
It should be cuBLAS's responsibility to check kernel correctness. We should focus on checking that we are dispatching correctly.
```suggestion
        # Full tile cases
        (128, 128),
        (256, 1024),
        (1024, 256),
        # Padded cases
        (256, 272),
        (303, 300),
```

kwyss-nvidia replied:
This is not checking the GEMM. It checks the quantizer kernel (in TE) versus a reference.

zhongbozhu commented on transformer_engine/pytorch/module/linear.py:
```diff
@@ -143,19 +148,25 @@ def forward(
+                else:
+                    if not isinstance(inputmat, QuantizedTensor):
+                        columnwise_usage = backward_needs_input and isinstance(
+                            input_quantizer, (MXFP8Quantizer, Float8BlockQuantizer)
```
Float8BlockQuantizer getting added here to set columnwise usage True

But this logic is also not doing anything unless we do fp8 gather, I am okay with keeping it there, just taking a note case in case we forget.

kwyss-nvidia replied:
PTAL. Added an assert.

timmoon10 commented:
/te-ci L1

kwyss-nvidia commented on transformer_engine/pytorch/distributed.py:
```diff
@@ -953,11 +953,15 @@ def _all_gather_fp8_blockwise(
-    the result, and this may be possible via calling `get_future()` on the
+    NOTE: The implementation is not sophisticated enough to honor async_op=True
+    and also apply the quantizer if quantizer=None. In such a case, it falls
+    back to a synchronous gather and invokes the quantizer.
```
Done.

kwyss-nvidia commented on transformer_engine/common/include/transformer_engine/cast.h:
```diff
@@ -102,6 +102,19 @@ void nvte_quantize(const NVTETensor input, NVTETensor output, cudaStream_t strea
                         cudaStream_t stream);
 
+/*! \brief Casts input tensor to FP8/MXFP8/BlockwiseFP8, providing the option to immediately exit the kernel
+ *         by configuring a noop in quant_config.
```
Done.

kwyss-nvidia commented on tests/pytorch/test_float8tensor.py:
```diff
@@ -132,6 +133,17 @@ def _test_quantize_dequantize(
+            else:
+                torch.testing.assert_close(x_fp8, x_ref_noop_test, **_tols[fp8_dtype])
+            return
+
```
Agreed. Pulled the noop branch into that test case.

yaox12 commented on transformer_engine/common/recipe/__init__.py:
```diff
@@ -287,3 +292,97 @@ def __post_init__(self) -> None:
+
+    In this strategy, tensors are scaled in blockwise fashion. Values within
+    each block share a common scaling factor. The block dimensionality
+    can be configured. The scaling factors are float32.
```
Maybe
```
The scaling factors are E8 by default. Switch to float32 by setting environment variable `NVTE_FP8_BLOCK_SCALING_FP32_SCALES=1`.
```

kwyss-nvidia replied:
Reworded, and I make the distinction between fp32 container types and whether the values are constrained to powers of 2, mentioning the env var. We never actually use e8m0 containers in this code.

timmoon10 commented on transformer_engine/pytorch/csrc/common.h:
```diff
@@ -168,12 +168,13 @@ class Float8BlockQuantizer : public Quantizer {
   // Which float8 type is used for q data.
   DType dtype;
 
- private:
```
It's a bit suspect how we're exposing the member variables publicly.

yaox12 replied:
We need to access `force_pow_2_scales` and `amax_epsilon` in https://github.com/NVIDIA/TransformerEngine/blob/59cb49c4cfdfc56fba52fcc617ccc8b15f1a9a0b/transformer_engine/pytorch/csrc/extensions/cast.cpp#L77-L78

Changed it to public members as in `Float8CurrentScalingQuantizer`.

timmoon10 commented on tests/pytorch/test_float8blockwisetensor.py:
```diff
@@ -111,6 +111,10 @@ def _test_quantize_dequantize(
 
         # Initialize random data
         x_ref = 2 * torch.rand(dims, dtype=dtype, device="cpu") - 1
+        if x_ref.numel() == 1:
+            # Reduce the chance that with trigger the trivial pass
+            # logic when all elements of x are close to zero.
+            x_ref.view(1)[0] = 42.1415
```
Slightly generalizing:
```suggestion
        # Initialize random data
        # Note: Make sure values are not all close to zero, or else test may pass trivially
        x_ref = 2 * torch.rand(dims, dtype=dtype, device="cpu") - 1
        x_ref.view(-1)[0] = 0.75
```

timmoon10 reviewed (approved):
LGTM aside from some nits

timmoon10 commented:
/te-ci L1

yaox12 commented:
/te-ci pytorch L1

yaox12 commented:
/te-ci pytorch

yaox12 commented:
JAX L0/L1 tests passed with build 26726340.
PyTorch L1 tests passed with build 26734338 except for irrelevant failures.
PyTorch L0 tests passed with build 26736222 except for irrelevant failures.

zhongbozhu commented:
/te-ci L1

zhongbozhu commented:
/te-ci pytorch L1
